drivers/net/loopback.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2015 Intel Corporation
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Network loopback interface implementation.
+ */
+
+#define SYS_LOG_DOMAIN "netlo"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_NETLO_LEVEL
+#include <logging/sys_log.h>
+
+#include <misc/printk.h>
+
+#include <net/net_pkt.h>
+#include <net/buf.h>
+#include <net/net_ip.h>
+#include <net/net_if.h>
+
+int loopback_dev_init(struct device *dev)
+{
+	return 0;
+}
+
+static void loopback_init(struct net_if *iface)
+{
+	net_if_set_link_addr(iface, "123456", 6, NET_LINK_DUMMY);
+}
+
+static int loopback_send(struct net_if *iface, struct net_pkt *pkt)
+{
+	int res;
+
+	if (!pkt->frags) {
+		SYS_LOG_ERR("No data to send");
+		return -ENODATA;
+	}
+
+	/* We need to swap the IP addresses because otherwise
+	 * the packet will be dropped.
+	 */
+
```

Comment:
Swapping IP address is probably not enough for UDP / TCP data, so the port numbers should be swapped too.

Reply:
It actually made me ponder what it means to implement a loopback interface. Based on its description (everything which gets send to it, gets received from it), it must swap IP addresses. But loopback interface is not an emulator of TCP/UDP echo servers. Like, you can't do on Linux:

~~~
nc -u 127.0.0.1 56687
~~~

and expect that data we input gets back to us. So, I guess swapping just addresses should be good enough.

Reply:
That is certainly true. We can leave it like this, it is just that depending on what one wants to do with this loopback interface, the application might need to know what is going on.

Reply:
Yeah, I'd suggest we start like that, and see where it leads us. An immediate usecase is just having an interface we can bind to, which doesn't have any external requirements to run (like specific hardware, a free UART, etc.)